### ISSUE TITLE
[FEATURE NAME] cr.pl.cluster_lineage throws MaybeEncodingError: Error sending result with many cells

### ISSUE THREAD

**Issue by nathanin** [theislab/cellrank#460]
`cr.pl.cluster_lineage` throws `MaybeEncodingError` when too many cells are used

```python
import cellrank as cr

# preprocessing / initial states / terminal states / lineages

model = cr.ul.models.GAM(adata)
cr.pl.cluster_lineage(
    adata,
    model,
    genes,
    lineage="Lineage 1",
    n_points=500,
    time_key="latent_time",
    show_progress_bar=True,
    n_jobs=12,
    recompute=True,
    sharey=True,
    backend='loky' # also 'multiprocessing' and 'threading'
)

```

When adata has > 10k cells:

```pytb
MaybeEncodingError: Error sending result: '[{'SLC6A14': {'Lineage 1': <GAM[gene='SLC6A14', lineage='Lineage 1', model=GammaGAM(callbacks=[Deviance(), Diffs()], fit_intercept=True, max_iter=2000, scale=None, terms=s(0) + intercept, tol=0.0001, verbose=False)]>}, 'AZGP1': {'Lineage 1': <GAM[gene='AZGP1', lineage='Lineage 1', model=GammaGAM(callbacks=[Deviance(), Diffs()], fit_intercept=True, max_iter=2000, scale=None, terms=s(0) + intercept, tol=0.0001, verbose=False)]>}, 'NMRAL1': {'Lineage 1': <GAM[gene='NMRAL1', lineage='Lineage 1', model=GammaGAM(callbacks=[Deviance(), Diffs()], fit_intercept=True, max_iter=2000, scale=None, terms=s(0) + intercept, tol=0.0001, verbose=False)]>}}]'. Reason: 'error("'i' format requires -2147483648 <= number <= 2147483647")
```

#### Versions:

> cellrank 1.1.0
scanpy 1.6.0
scvelo 0.2.2
anndata 0.7.5
joblib 1.0.0

------

I'm testing around this issue, and it looks like the upper limit of cells is somewhere between 7,500 - 10,000 (7,500 works, 10k fails with the above error). It could be related to this issue with joblib: https://github.com/joblib/joblib/issues/818 and the discussion there suggests to use a  different parallel backend, however I'm seeing the same issue using both `multiprocessing` and `threading`. For now I'm getting through by subsetting my dataset. 

Another note, I'm performing the subsetting according to terminal state probability -- taking the highest scoring cells to fit the gene trend models. Does this sound like a reasonable approach?

[environment.yml.txt](https://github.com/theislab/cellrank/files/5805128/environment.yml.txt)

**Comment by michalk8** [comment]
Hi @nathanin ,
thanks for posting this, I think I've encountered this bug before, though not in this context.
Based on the env you've provided, it looks like you're using Python3.7, is that right? This bug appears to have been already fixed in https://github.com/python/cpython/pull/10305
I'm using Python3.8 and the changes are already there - could you try updating to 3.8 and see if it persists?

Regarding the subsetting of high-likely cells - what sort of a threshold are you using?
During the model's preparation, something similar happens - the weights (absorption probabilities) are modified such that any cell with a weight (absorption probability) below some threshold (default is `0.01`) is set to that threshold [here](https://cellrank.readthedocs.io/en/latest/classes.html#cellrank.ul.models.BaseModel.prepare).
I wouldnt't filter out the cells unless you're specifically interested in those high-likely cells or the fit is quite bad in the beginning (can happen if there are cells which are extremely unlikely to develop towards an endpoint early-on). Maybe @Marius1311 can also weigh-in on this.

**Comment by nathanin** [comment]
Yep, in an environment built on python 3.8 I can run all my cells (~43k) through `cluster_lineage` at once. I guess this renders my second question moot as it's preferable to use all cells. In any case I was choosing cells to subset according to terminal state scores this way:

```python
N = 5000
top_cells = pd.DataFrame(ad.obsm['to_terminal_states']['Lineage 1'], 
                            index=ad.obs_names,
                            columns=['terminal_state_score'])
top_cells = top_cells.sort_values('terminal_state_score', ascending=False)
chosen_cells = top_cells.index[:N]
```

Thank you for getting back so quickly! I'm going to mark this issue closed as upgrading to python 3.8 solved the core issue.

**Comment by Marius1311** [comment]
Hi @nathanin and thanks at @michalk8 for already solving the core issue. Just a comment from my side, I would in fact use all cells as the fate probabilities should already take care of the relative weighting of cells with respect to lineages.
